### ISSUE TITLE
A general query

### ISSUE THREAD

**Issue by bamdevm** [pymanopt/pymanopt#120]
Hello,
How does auto differentiation for a loss function work which itself is an outcome of an iterative process? 

In particular, loss(x, a) = min_z f(z, x, a), where the min_z f can be solved with a gradient descent algorithm that can be coded. I want to compute the gradient of the loss with respect to x.  Does auto differentiation work for this?

Regards

**Comment by mshvartsman** [comment]
Technically if the solver itself is differentiable you could do something like this (see e.g. https://arxiv.org/abs/1910.12430) but I think this question is unrelated to pymanopt.

**Comment by bamdevm** [comment]
Thanks. 

I am aware that the question is general in nature, and hence the particular title. 

My main interest was get some opinions on the ability of auto differentiation to take care of minmax forms. More concretely, I have a problem 

min_{x in manifold} max_{z in convex set}  f(z,x).

Given x, I can solve the max problem with an iterative solver. However, for the min problem, I would be required to compute the gradient which I intend to do with autograd.  

Will it work? If yes, how computationally efficient would it be?

Regards,

**Comment by NicolasBoumal** [comment]
Hello Bamdev,

I believe that the way AD works is: it would implicitly "unroll" any loops
(for, while) so that the inner algorithm can be seen as a function with a
certain well defined sequence of operations, and it would differentiate
through that. If this 'unrolled' function isn't too long and, crucially, if
the operations in it (including the stopping criterion!) do not depend on
the results of the intermediate computations, then I'd expect AD to succeed
-- possibly slowly. However, if your inner solver for the convex problem
has a stopping criterion that depends on the intermediate iterates to
decide when to stop, then I'd expect AD to fail, because the unrolled
function wouldn't be differentiable (I imagine).

Just my two cents: I don't know much about AD.

Nicolas

**Comment by bamdevm** [comment]
Got it, Nicolas.

Thanks.

**Comment by Lezcano** [comment]
This issue is somewhat old, but in case it is still of interest:

Autodiff on PyTorch (in Tensorflow is _much_ trickier, unless one uses eager execution) is able to backpropagate through sequences of arbitrary length. Even when one has a stopping time that might depend on the input or a random number. The way it works is via what's called a "tape". When $f(x)$ is computed, the tape records the intermediate operations that were executed to get to the value of $f(x)$, then the gradients are simply computed by rolling back this dependency tree.

That being said, one has to be careful when letting the autodiff engine doing all the heavy-lifting. I particular, it is clear that $C^0$ convergence does not imply $C^1$ convergence. As such, if you have an algorithm to approximate the value of a function (say f(x) = \min g(x,z)), it does not mean that the derivative of the algorithm will approximate the derivative of the function. One may prove that that is the case in specific cases, but it is far from true in general. As such, sometimes it is necessary to compute analytically the (sub-)gradient of the function (or come up with an approximation algorithm) and implement it by hand.

**Comment by NicolasBoumal** [comment]
Recent paper on this topic:
https://arxiv.org/abs/2002.03722
Super-efficiency of automatic differentiation for functions defined as a minimum
Pierre Ablin, Gabriel Peyré, Thomas Moreau

**Comment by sweichwald** [comment]
What @bamdevm described may also be a setting where a set-up with jax may make things more intuitive (and easy) to set-up :bulb:
We do not (yet) have a jax backend, though.

**Comment by nkoep** [comment]
I believe what @Lezcano was implying with his comment regarding tensorflow relates to tensorflow 1.x (as well as theano). Since these build a computational graph first it's much harder to define arbitrary computations that require control flow structures. After adding support for TF2 (and dropping support for TF1) this is less of an issue, since TF2 (as well as autograd and pytorch) use tape-based autodiff which simply records the operations that were actually applied to the input to compute gradients in reverse mode. That's why tape-based systems have no issue with arbitrarily complex control flow.

I'll close this topic for now. Feel free to re-open it again as needed :pray:
